# **MovieLens 1M Dataset Analysis**
# **(MovieLens 1M Veri Kümesi Analizi)**

GroupLens Research has collected and made available rating data sets from the MovieLens web site (https://grouplens.org/datasets/movielens/). 
<br> The data sets were collected over various periods of time, depending on the size of the set.
<br> MovieLens 1M Dataset: collected from users of MovieLens in the late 1990s and early 2000s.
<br> 1 million ratings from 6000 users on 4000 movies. 
<br> It’s spread across 3 tables: ratings, user information, and movie information. 
<br> You can find **users.dat , movies.dat , ratings.dat** file in the ml-1m folder.

GroupLens Research **,** MovieLens web sitesinden (https://grouplens.org/datasets/movielens/) derecelendirme (rating) veri setleri toplamış ve kullanıma sunmuştur.
<br> Veri setleri **,** setin büyüklüğüne bağlı olarak çeşitli zaman dilimlerinde toplanmıştır.
<br> **MovieLens 1M Veri Kümesi : 1990**'ların sonunda ve **2000**'lerin başında **MovieLens** kullanıcılarından toplanmıştır. **( 4000** filmde **6000** kullanıcıdan **1** milyon oy **).**
<br> **3** tabloya yayılmıştır **: kullanıcı bilgileri , film bilgileri** ve **derecelendirmeler(rating).**
<br> **'users.dat' , 'movies.dat' , 'ratings.dat'** dosyalarını **, '``Movielens-1m``' klasöründe** bulabilirsiniz...

## First, read these data files into a dataframe object and merge them **:**
## (İlk olarak **,** bu veri dosyalarını bir dataframe nesnesine okuyun ve birleştirin) **:**

**++++ NOTE ++++**
+ Buradaki dosyamızı (veri setimizi) illaki **.read_table( )** fonksiyonu ile okumamıza gerek yok. **.read_csv( )** fonksiyonu ile de okuyabilirdik...
+ Okunmaya çalışılan dosyanın (veri setinin) içerisindeki veriler **,** birbirlerinden nasıl ayrıldılarsa **, '(sep=...)'** değeri de ona göre belirlenir.
+ Pandas normalde veri setindeki ilk satırı başlık (Kolon isimleri) olarak görür. Eğer veri setimizde başlık yazılmamış **,** direkt verilerin içeriğine geçilmişse **'(header=None)'** komudunu yazmamız gerekir. Yani bu veri setinde başlık yok dememiz gerekir.
+ Daha sonra da başlığımızı **,** yani Kolon isimlerini belirlememiz için **'(names=...)'** komutunu kullanmamız gerekir.
+ **'(engine=...)'** komutunu da **;** nasıl okumamız **,** ne ile okumamız gerektiğini belirtmek için kullanırız.

In [1]:
import pandas as pd

+ Öncelikle **, 'users.dat'** Veri Setini çağıralım **:**

In [2]:
unames = ["user_id" , "gender" , "age" , "occupation" , "zip"]
users = pd.read_table("users.dat" , sep="::" , header=None , names=unames , engine="python")

In [3]:
users.head()

user_id gender  age  occupation    zip
0        1      F    1          10  48067
1        2      M   56          16  70072
2        3      M   25          15  55117
3        4      M   45           7  02460
4        5      M   25          20  55455

+ **'ratings.dat'** Veri Seti çok büyük olduğu için çalıştırılması biraz zaman alabilir **:**

In [5]:
rnames = ["user_id" , "movie_id" , "rating" , "timestamp"]
ratings = pd.read_table("ratings.dat" , sep="::" , header=None , names=rnames , engine="python")

In [4]:
ratings.head()

user_id  movie_id  rating  timestamp
0        1      1193       5  978300760
1        1       661       3  978302109
2        1       914       3  978301968
3        1      3408       4  978300275
4        1      2355       5  978824291

Yukarıdaki veri setimizdeki ilk beş verimizin arasından **,** ilk veriye baktığımızda **; 1 numaralı kullanıcı , 1193 numaralı filme 5 puan vermiş** demek istiyor. **'timestamp'** Kolonu da bu puanın verildiği tarihi gösteriyor. Tarih **,** dikkat ettiysek farklı bir şekilde yazılmış... İleride öğreneceğiz...

**++++ NOTE ++++**
<br> Eğer aşağıda **, 'movies.dat'** dosyasını okurken **, (encoding='latin-1')** komutunu yazmasaydık **'utf-8'** hatası alacaktık...
<br> Hatanın Tam Yazılışı **::** 'utf-8' codec can't decode byte 0xe9 in position 3114: invalid continuation byte.

In [6]:
mnames = ["movie_id" , "title" , "genres"]
movies = pd.read_table("movies.dat" , sep="::" , header=None , names=mnames , engine="python" , encoding='latin-1')

In [6]:
movies.head()

movie_id                               title                        genres
0         1                    Toy Story (1995)   Animation|Children's|Comedy
1         2                      Jumanji (1995)  Adventure|Children's|Fantasy
2         3             Grumpier Old Men (1995)                Comedy|Romance
3         4            Waiting to Exhale (1995)                  Comedy|Drama
4         5  Father of the Bride Part II (1995)                        Comedy

Şimdi bu yukarıdaki **3 farklı** veri seti tablosunu **birleştirmemiz** gerekiyor. Sebebi de **;** analiz yapmak için bu üç tablodaki verileri de bir arada kullanmamız gerekiyor. Mesela diyelim ki **,** cinsiyete bağlı bir analiz yapmak istiyoruz. Kadınların en çok beğendiği film türünü bulmaya çalışacağız. Böyle bir durumda bu üç tabloyu da bir arada kullanmamız gerekecek...

## Make one dataframe by merging three dataframes **:**
## (Üç veri çerçevesini (dataFrame) birleştirerek bir veri çerçevesi yapın **:**)
+ [ **``pd.merge( )``** fonksiyonu ile **:** ]
+ **MERGE** fonksiyonu ile yapacağımız için önce iki tane tabloyu **ortak kolonlara (key'lere) göre birleştirmeliyiz.** Sonra da **,** bu yeni oluşan tablo ile diğer tabloyu yine ortak kolonlara (key'lere) göre birleştirmeliyiz **:**

**+--- NOTE --->** Ortak Kolonlar birer tane olduğu için **; '(on=...)' , '(left_on=...)' , '(right_on=...)'** gibi komutlar ile hangi Kolona göre birleştirme işlemi yapacağımızı yazmamıza gerek yok...

+ Yine bildiğimiz gibi **'pd.merge( )'** fonksiyonunda DEFAULT olarak **'(how=inner)'** komutu çalışıyor... Yani birleştirilen tabloların ikisinde birden ortak olarak BULUNMAYAN veriyi **,** yeni oluşturduğumuz tabloya ALMIYORDU... Peki mesela **'users'** ve **'ratings'** tablolarının ortak kolonları olan **'user_id'** kolonunda **,** bu iki tablo için **,** ortak olarak BULUNMAYAN veriler varsa **,** bunları nasıl görebiliriz **?**

**++++ NOTE ++++**
<br> **``set( )``** fonksiyonu **,** KÜMELERİ ifade eder...
+ Aşağıda da öncelikle her iki tablonun da **'user_id'** Kolonları için **unique** olan verileri bulup **,** sonra da bunları Küme içerisine alıp **,** bir kümeden diğer kümeyi çıkartarak **,** Ortak Olmayan verileri bulmaya çalışacağız... **:**

In [8]:
set(users.user_id.unique() - ratings.user_id.unique())

{0}

+ Bir de tam tersi şekilde bir çıkartma işlemi yapalım **:**

In [9]:
set(ratings.user_id.unique() - users.user_id.unique())

{0}

+ Hiç **,** ortak olmayan veri yokmuş gibi görünüyor... Bir de eleman sayılarına bakalım. Gerçekten aynı sayıda veriye mi sahipler **?**

In [10]:
ratings.user_id.nunique()

6040

In [11]:
users.user_id.nunique()

6040

**Evet yukarıda gördüğümüz gibi, gerçekten de her iki tabloda da ortak olan _"user_id"_ kolonlarında, ortak OLMAYAN hiçbir veri yokmuş...**
+ Şimdi tekrar kaldığımız yerden **,** tabloları birleştirme işlemine devam edelim **:**

In [7]:
binary_data = pd.merge(users , ratings)
binary_data.head()

user_id gender  age  occupation    zip  movie_id  rating  timestamp
0        1      F    1          10  48067      1193       5  978300760
1        1      F    1          10  48067       661       3  978302109
2        1      F    1          10  48067       914       3  978301968
3        1      F    1          10  48067      3408       4  978300275
4        1      F    1          10  48067      2355       5  978824291

In [8]:
data = pd.merge(binary_data , movies)
data.head()

user_id gender  age  occupation    zip  movie_id  rating  timestamp  \
0        1      F    1          10  48067      1193       5  978300760   
1        2      M   56          16  70072      1193       5  978298413   
2       12      M   25          12  32793      1193       4  978220179   
3       15      M   25           7  22903      1193       4  978199279   
4       17      M   50           1  95350      1193       5  978158471   

                                    title genres  
0  One Flew Over the Cuckoo's Nest (1975)  Drama  
1  One Flew Over the Cuckoo's Nest (1975)  Drama  
2  One Flew Over the Cuckoo's Nest (1975)  Drama  
3  One Flew Over the Cuckoo's Nest (1975)  Drama  
4  One Flew Over the Cuckoo's Nest (1975)  Drama

+ Yukarıdaki gibi ayrı ayrı iki farklı KOD satırı yazmadan **,** tek KOD satırında şu şekilde yapabiliriz **:**

In [7]:
data = pd.merge(pd.merge(users , ratings) , movies)
data.head()

user_id gender  age  occupation    zip  movie_id  rating  timestamp  \
0        1      F    1          10  48067      1193       5  978300760   
1        2      M   56          16  70072      1193       5  978298413   
2       12      M   25          12  32793      1193       4  978220179   
3       15      M   25           7  22903      1193       4  978199279   
4       17      M   50           1  95350      1193       5  978158471   

                                    title genres  
0  One Flew Over the Cuckoo's Nest (1975)  Drama  
1  One Flew Over the Cuckoo's Nest (1975)  Drama  
2  One Flew Over the Cuckoo's Nest (1975)  Drama  
3  One Flew Over the Cuckoo's Nest (1975)  Drama  
4  One Flew Over the Cuckoo's Nest (1975)  Drama

**################################################**
<br> **---- occupation ==** Meslek **,** İş **,** Uğraş.
<br> **---- time stamp ==** Tarih bilgisi **,** Zaman bilgisi **,** Tarih damgası.
<br> **---- genres ==** Film türleri.
<br> **################################################**

+ Şimdi öncelikle tablolar birleştirilmeden önce **,** toplam kaç tane veri (eleman **,** hücre) varmış ona bakalım **:**

In [15]:
users.size + ratings.size + movies.size

4042685

+ Şimdi de bu tablomuzun **,** bu son birleştirilmiş halinde toplam kaç tane veri (eleman **,** hücre) var bakalım **:**

In [16]:
data.size

10002090

Yani yukarıda gördüğümüz gibi **,** tabloların ayrı ayrı şekilde hazırlanmasının sebebi **,** çok daha küçük boyutlarda veri setleri hazırlayabilmemizdendir... Hepsi bir arada tek tabloda olsaydı **,** ayrı ayrı hepsinin toplamından çok daha büyük bir veri seti ortaya çıkacaktı...

**+--- NOTE --->** Yani bu veri setlerini bizimle paylaşan kuruluşun **,** her bir veri setini (dosya **,** tablo) ayrı ayrı olarak değil de hepsi tek bir veri seti olarak AYARLAMAMASININ sebebi **;** hepsi bir arada tek veri seti şeklinde olduğunda veri setinin boyutu çok büyük hale geliyor. Böyle bir durum yaşanmaması için de bu şekilde üç farklı veri seti halinde hazırlanmış...

**+--- NOTE ---> MERGE** ile değil de **JOİN** ile yapmaya çalışırsak da hata alırız. Çünkü bu üç tabloda ortak kolonlar (keyler) var. Yani aynı isimde kolonlar var. Ve bu şekilde ortak kolonları olan tabloları yan yana birleştiremeyiz.
<br> Ayrıca **,** aynı kolonları tekrar tekrar yan yana birleştirmeye çalışmak gibi anlamsız bir durum ortaya çıkar... 

### Calculate mean ratings for each movie with gender difference **:**
### (Cinsiyete göre **,** her film için ortalama puanları hesaplayın **:**)

+ Burada **,** her film için ORTALAMA **rating**'i hesaplayalım. Ve bunu **,** kadınlar için ayrı erkekler için ayrı hesaplayalım **:**
+ Ayrıca bunu **'``pivot_table()``'** metodu ile yapalım **:**
+ Her film için Ortalama Raitingleri hesaplamak istediğimiz için **, 'pivot_table( )'**'nin **values** değeri haliyle **'rating'** Kolonu olacak... **Gruplamayı** da iki farklı Kolona göre yapmamız daha mantıklı olur **; Cinsiyet (gender)** ve **Film ismi (title).** Bu gruplamada da **;** mesela film isimlerini **index** olarak gösterip **,** cinsiyetleri de kadın - erkek şeklinde **kolonda** göstersek çok daha güzel bir gösterim olur **:** 

In [17]:
data.head()

user_id gender  age  occupation    zip  movie_id  rating  timestamp  \
0        1      F    1          10  48067      1193       5  978300760   
1        2      M   56          16  70072      1193       5  978298413   
2       12      M   25          12  32793      1193       4  978220179   
3       15      M   25           7  22903      1193       4  978199279   
4       17      M   50           1  95350      1193       5  978158471   

                                    title genres  
0  One Flew Over the Cuckoo's Nest (1975)  Drama  
1  One Flew Over the Cuckoo's Nest (1975)  Drama  
2  One Flew Over the Cuckoo's Nest (1975)  Drama  
3  One Flew Over the Cuckoo's Nest (1975)  Drama  
4  One Flew Over the Cuckoo's Nest (1975)  Drama

**+--- NOTE ---> pivot_table( )** metodunun içerisindeki DEFAULT fonksiyon zaten ortalamayı hesaplayan **'mean'** fonksiyonudur. Bunu değiştirmemize gerek yok... (**SHIFT+TAB** tuşlarına basarak da görebiliriz...) **:**

In [8]:
mean_ratings = data.pivot_table(values='rating' , index='title' , columns='gender')

In [11]:
mean_ratings.head()

gender                                F         M
title                                            
$1,000,000 Duck (1971)         3.375000  2.761905
'Night Mother (1986)           3.388889  3.352941
'Til There Was You (1997)      2.675676  2.733333
'burbs, The (1989)             2.793478  2.962085
...And Justice for All (1979)  3.828571  3.689024

### Let’s  filter down to movies that received at least 250 ratings (a completely arbitrary number) **:**
### (**En az 250** reyting alan (tamamen rastgele bir sayı) filmleri filtreleyelim **:** )

In [25]:
data.head()

user_id gender  age  occupation    zip  movie_id  rating  timestamp  \
0        1      F    1          10  48067      1193       5  978300760   
1        2      M   56          16  70072      1193       5  978298413   
2       12      M   25          12  32793      1193       4  978220179   
3       15      M   25           7  22903      1193       4  978199279   
4       17      M   50           1  95350      1193       5  978158471   

                                    title genres  
0  One Flew Over the Cuckoo's Nest (1975)  Drama  
1  One Flew Over the Cuckoo's Nest (1975)  Drama  
2  One Flew Over the Cuckoo's Nest (1975)  Drama  
3  One Flew Over the Cuckoo's Nest (1975)  Drama  
4  One Flew Over the Cuckoo's Nest (1975)  Drama

+ Yani belli bir sayıdan daha az oylanan filmleri hesaba katmak istemiyoruz... **:**
+ İlk olarak **, her bir** filmin kaç tane **'rating'** verisi var **?** Bunu hesaplayalım **:**
+ Bunu **``.groupby( )``** metodu veya **``.value_counts( )``** metodu ile yapabiliriz... **:**

In [9]:
ratings_by_title = data.groupby('title').count().sort_values(by='rating' , ascending=False)

In [13]:
ratings_by_title.head()

user_id  gender   age  \
title                                                                       
American Beauty (1999)                                 3428    3428  3428   
Star Wars: Episode IV - A New Hope (1977)              2991    2991  2991   
Star Wars: Episode V - The Empire Strikes Back ...     2990    2990  2990   
Star Wars: Episode VI - Return of the Jedi (1983)      2883    2883  2883   
Jurassic Park (1993)                                   2672    2672  2672   

                                                    occupation   zip  \
title                                                                  
American Beauty (1999)                                    3428  3428   
Star Wars: Episode IV - A New Hope (1977)                 2991  2991   
Star Wars: Episode V - The Empire Strikes Back ...        2990  2990   
Star Wars: Episode VI - Return of the Jedi (1983)         2883  2883   
Jurassic Park (1993)                                      2672  2672   

                                                    movie_id  rating  \
title                                                                  
American Beauty (1999)                                  3428    3428   
Star Wars: Episode IV - A New Hope (1977)               2991    2991   
Star Wars: Episode V - The Empire Strikes Back ...      2990    2990   
Star Wars: Episode VI - Return of the Jedi (1983)       2883    2883   
Jurassic Park (1993)                                    2672    2672   

                                                    timestamp  genres  
title                                                                  
American Beauty (1999)                                   3428    3428  
Star Wars: Episode IV - A New Hope (1977)                2991    2991  
Star Wars: Episode V - The Empire Strikes Back ...       2990    2990  
Star Wars: Episode VI - Return of the Jedi (1983)        2883    2883  
Jurassic Park (1993)                                     2672    2672

In [10]:
#   OR

ratings_by_title = data.groupby('title').size().sort_values(ascending = False) ;
ratings_by_title

title
American Beauty (1999)                                       3428
Star Wars: Episode IV - A New Hope (1977)                    2991
Star Wars: Episode V - The Empire Strikes Back (1980)        2990
Star Wars: Episode VI - Return of the Jedi (1983)            2883
Jurassic Park (1993)                                         2672
                                                             ... 
Target (1995)                                                   1
I Don't Want to Talk About It (De eso no se habla) (1993)       1
An Unforgettable Summer (1994)                                  1
Never Met Picasso (1996)                                        1
Full Speed (1996)                                               1
Length: 3706, dtype: int64

**++++ NOTE ++++**
<br> Bu yukarıdaki **.groupby( ).size( )** metodu da **; groupby( )** fonksiyonunun içerisine yazdığımız Kolonu **,** Tablolara bölüyor. Yani burada her film için Ayrı Bir Tablo yaratıyor...
<br> **.size( )** metodu ile de her tablonun ayrı ayrı **,** kaç satırı var onu görüyoruz... AMA bu metod **,** Satırda bulunan **NaN** değerleri de **,** Satırda olduğu için sayar!!!

In [15]:
#   OR

ratings_by_title = data.value_counts('title') ;
ratings_by_title

title
American Beauty (1999)                                       3428
Star Wars: Episode IV - A New Hope (1977)                    2991
Star Wars: Episode V - The Empire Strikes Back (1980)        2990
Star Wars: Episode VI - Return of the Jedi (1983)            2883
Jurassic Park (1993)                                         2672
                                                             ... 
Target (1995)                                                   1
I Don't Want to Talk About It (De eso no se habla) (1993)       1
An Unforgettable Summer (1994)                                  1
Never Met Picasso (1996)                                        1
Full Speed (1996)                                               1
Length: 3706, dtype: int64

**++++ NOTE ++++**
<br> **".value_counts( )"** metodu kullanırken dikkat etmemiz gereken şey de **;** bu metod da **NaN** değerleri sayar!!! Yani **NaN** olan değerleri de toplama işlemine dahil eder!!!
<br> **ANCAK , ".groupby( ).count( )"** metodu ile sayma işlemini yaparsak **,** bu metod **NaN** değerleri sayma işlemine dahil etmez!!!

+ **Get titles that have at least 250 ratings :**
<br> **En az 250 reytinge sahip başlıkları alın :**

In [11]:
ratings_by_title[ratings_by_title >= 250]

title
American Beauty (1999)                                            3428
Star Wars: Episode IV - A New Hope (1977)                         2991
Star Wars: Episode V - The Empire Strikes Back (1980)             2990
Star Wars: Episode VI - Return of the Jedi (1983)                 2883
Jurassic Park (1993)                                              2672
                                                                  ... 
Presidio, The (1988)                                               251
Top Hat (1935)                                                     251
Teenage Mutant Ninja Turtles II: The Secret of the Ooze (1991)     251
U Turn (1997)                                                      250
Random Hearts (1999)                                               250
Length: 1216, dtype: int64

+ Yani **250'den fazla** rating alan filmlerin SADECE İSİMLERİNİ görmek istiyoruz **:**
<br> Bunun için de **, "``.index``"** metodunu kullanacağız **:**

In [12]:
active_titles = ratings_by_title[ratings_by_title >= 250].index
active_titles

Index(['American Beauty (1999)', 'Star Wars: Episode IV - A New Hope (1977)',
       'Star Wars: Episode V - The Empire Strikes Back (1980)',
       'Star Wars: Episode VI - Return of the Jedi (1983)',
       'Jurassic Park (1993)', 'Saving Private Ryan (1998)',
       'Terminator 2: Judgment Day (1991)', 'Matrix, The (1999)',
       'Back to the Future (1985)', 'Silence of the Lambs, The (1991)',
       ...
       'Bicycle Thief, The (Ladri di biciclette) (1948)',
       'Take the Money and Run (1969)', 'Bachelor, The (1999)',
       'Man Who Knew Too Little, The (1997)',
       'Wes Craven's New Nightmare (1994)', 'Presidio, The (1988)',
       'Top Hat (1935)',
       'Teenage Mutant Ninja Turtles II: The Secret of the Ooze (1991)',
       'U Turn (1997)', 'Random Hearts (1999)'],
      dtype='object', name='title', length=1216)

+ Get mean ratings for these **active titles :**
<br> Bu **active_titles** ataması için Ortalama Reytingleri alın **:**
+ Yani daha yukarıda **mean_ratings** ataması ile **, kadın - erkek** ayrımına göre Ortalama Reytingleri alınan filmlerin arasından **,** sadece 250'den fazla rating alan filmleri çekmek istiyoruz **:**
<br> Yani **mean_ratings** tablomuzun içinden **,** sadece **active_titles** tablomuzda yazan INDEXLERE erişmek istiyoruz..!
<br> Yani Indexe taşıdığımız **'title'**'ye erişmek istiyoruz..! Bunu da **``.loc[ ]``** fonksiyonu ile yapabiliriz... Çünkü INDEXTEN bir şey aratmak için **, .loc[ ]** fonksiyonunu kullanırız... **:**

In [17]:
mean_ratings.head()

gender                                F         M
title                                            
$1,000,000 Duck (1971)         3.375000  2.761905
'Night Mother (1986)           3.388889  3.352941
'Til There Was You (1997)      2.675676  2.733333
'burbs, The (1989)             2.793478  2.962085
...And Justice for All (1979)  3.828571  3.689024

In [19]:
active_titles

Index(['American Beauty (1999)', 'Star Wars: Episode IV - A New Hope (1977)',
       'Star Wars: Episode V - The Empire Strikes Back (1980)',
       'Star Wars: Episode VI - Return of the Jedi (1983)',
       'Jurassic Park (1993)', 'Saving Private Ryan (1998)',
       'Terminator 2: Judgment Day (1991)', 'Matrix, The (1999)',
       'Back to the Future (1985)', 'Silence of the Lambs, The (1991)',
       ...
       'Bicycle Thief, The (Ladri di biciclette) (1948)',
       'Take the Money and Run (1969)', 'Bachelor, The (1999)',
       'Man Who Knew Too Little, The (1997)',
       'Wes Craven's New Nightmare (1994)', 'Presidio, The (1988)',
       'Top Hat (1935)',
       'Teenage Mutant Ninja Turtles II: The Secret of the Ooze (1991)',
       'U Turn (1997)', 'Random Hearts (1999)'],
      dtype='object', name='title', length=1216)

In [13]:
mean_ratings_active_titles = mean_ratings.loc[active_titles]
mean_ratings_active_titles

gender                                                     F         M
title                                                                 
American Beauty (1999)                              4.238901  4.347301
Star Wars: Episode IV - A New Hope (1977)           4.302937  4.495307
Star Wars: Episode V - The Empire Strikes Back ...  4.106481  4.344577
Star Wars: Episode VI - Return of the Jedi (1983)   3.865237  4.069058
Jurassic Park (1993)                                3.579407  3.814197
...                                                      ...       ...
Presidio, The (1988)                                3.547619  3.253589
Top Hat (1935)                                      4.245763  4.060150
Teenage Mutant Ninja Turtles II: The Secret of ...  2.152174  2.121951
U Turn (1997)                                       2.931034  3.095023
Random Hearts (1999)                                2.483146  2.242236

[1216 rows x 2 columns]

**++++ NOTE ++++**
<br> Yukarıda **,** aradığımız şey INDEXTE değil de KOLONDA olsaydı **; ``.isin( )``** veya **``.query( )``** fonksiyonlarını kullanabilirdik...

+ Let's see the top movies amongst the female viewers **:**
<br> **Kadın** izleyiciler arasında en çok izlenen filmleri görelim **:**
<br> Yani **Kadın Kolonuna** göre **büyükten küçüğe** bir **sıralama** işlemi yapalım **:**

In [14]:
top_female_ratings = mean_ratings_active_titles.sort_values(by='F' , ascending=False);
top_female_ratings.head(10)

gender                                                     F         M
title                                                                 
Close Shave, A (1995)                               4.644444  4.473795
Wrong Trousers, The (1993)                          4.588235  4.478261
Sunset Blvd. (a.k.a. Sunset Boulevard) (1950)       4.572650  4.464589
Wallace & Gromit: The Best of Aardman Animation...  4.563107  4.385075
Schindler's List (1993)                             4.562602  4.491415
Shawshank Redemption, The (1994)                    4.539075  4.560625
Grand Day Out, A (1992)                             4.537879  4.293255
To Kill a Mockingbird (1962)                        4.536667  4.372611
Creature Comforts (1990)                            4.513889  4.272277
Usual Suspects, The (1995)                          4.513317  4.518248

**+--- NOTE --->** Yukarıda **'.sort_values( )'** metodunun içinde **SHIFT+TAB** yaparak **,** içerisindeki Parametrelerin nasıl kullanılabileceğini görebiliriz...

### **Measuring Rating Disagreement**
+ Suppose you wanted to find the movies that are most divisive between male and female viewers. 
<br>One way is to add a column to **'mean_ratings'** containing the difference in means, then sort by that **:** 
+ Erkek ve kadın izleyiciler arasında **,** oy farkının en fazla olduğu filmleri bulmak istediğinizi varsayalım.
<br> Bunun bir yolu **,** Ortalamalardaki farkı içeren **'mean_ratings'** Tablosuna bir **"difference" (fark)** Sütunu eklemek ve ardından buna göre sıralamaktır. **(** Eksili değerler de çıkabileceği için en son da **Mutlak Değer**'lerini alalım... **) :**

**+--- NOTE ---> Mutlak Değer** almak için de **,** "**``abs( )``**" fonksiyonunu kullanırız **!**

In [17]:
mean_ratings_active_titles.head()

gender                                                     F         M
title                                                                 
American Beauty (1999)                              4.238901  4.347301
Star Wars: Episode IV - A New Hope (1977)           4.302937  4.495307
Star Wars: Episode V - The Empire Strikes Back ...  4.106481  4.344577
Star Wars: Episode VI - Return of the Jedi (1983)   3.865237  4.069058
Jurassic Park (1993)                                3.579407  3.814197

In [15]:
#   Add diff column to "mean_ratings_active_titles" :
#   "ortalama_ratings_active_titles"'a "different" sütunu ekle :

#   Take the absolute value of the "different" column :
#   Yarattığımız "different" sütununun , mutlak değerini al :

mean_ratings_active_titles["different"] = abs(mean_ratings_active_titles["F"] - mean_ratings_active_titles["M"]) ;
mean_ratings_active_titles

gender                                                     F         M  \
title                                                                    
American Beauty (1999)                              4.238901  4.347301   
Star Wars: Episode IV - A New Hope (1977)           4.302937  4.495307   
Star Wars: Episode V - The Empire Strikes Back ...  4.106481  4.344577   
Star Wars: Episode VI - Return of the Jedi (1983)   3.865237  4.069058   
Jurassic Park (1993)                                3.579407  3.814197   
...                                                      ...       ...   
Presidio, The (1988)                                3.547619  3.253589   
Top Hat (1935)                                      4.245763  4.060150   
Teenage Mutant Ninja Turtles II: The Secret of ...  2.152174  2.121951   
U Turn (1997)                                       2.931034  3.095023   
Random Hearts (1999)                                2.483146  2.242236   

gender                                              different  
title                                                          
American Beauty (1999)                               0.108400  
Star Wars: Episode IV - A New Hope (1977)            0.192371  
Star Wars: Episode V - The Empire Strikes Back ...   0.238096  
Star Wars: Episode VI - Return of the Jedi (1983)    0.203821  
Jurassic Park (1993)                                 0.234791  
...                                                       ...  
Presidio, The (1988)                                 0.294031  
Top Hat (1935)                                       0.185612  
Teenage Mutant Ninja Turtles II: The Secret of ...   0.030223  
U Turn (1997)                                        0.163988  
Random Hearts (1999)                                 0.240910  

[1216 rows x 3 columns]

In [20]:
#   OR

mean_ratings_active_titles["different"] = abs(mean_ratings_active_titles.F - mean_ratings_active_titles.M) ;
mean_ratings_active_titles

gender                                                     F         M  \
title                                                                    
American Beauty (1999)                              4.238901  4.347301   
Star Wars: Episode IV - A New Hope (1977)           4.302937  4.495307   
Star Wars: Episode V - The Empire Strikes Back ...  4.106481  4.344577   
Star Wars: Episode VI - Return of the Jedi (1983)   3.865237  4.069058   
Jurassic Park (1993)                                3.579407  3.814197   
...                                                      ...       ...   
Presidio, The (1988)                                3.547619  3.253589   
Top Hat (1935)                                      4.245763  4.060150   
Teenage Mutant Ninja Turtles II: The Secret of ...  2.152174  2.121951   
U Turn (1997)                                       2.931034  3.095023   
Random Hearts (1999)                                2.483146  2.242236   

gender                                              different  
title                                                          
American Beauty (1999)                               0.108400  
Star Wars: Episode IV - A New Hope (1977)            0.192371  
Star Wars: Episode V - The Empire Strikes Back ...   0.238096  
Star Wars: Episode VI - Return of the Jedi (1983)    0.203821  
Jurassic Park (1993)                                 0.234791  
...                                                       ...  
Presidio, The (1988)                                 0.294031  
Top Hat (1935)                                       0.185612  
Teenage Mutant Ninja Turtles II: The Secret of ...   0.030223  
U Turn (1997)                                        0.163988  
Random Hearts (1999)                                 0.240910  

[1216 rows x 3 columns]

In [16]:
#   Sort values by "different" column:
#   Değerleri , "different" sütununa göre sırala :
#   Bu bize , kadınlar ve erkekler arasındaki beğeni farkına göre bir sıralama yapacak :

mean_ratings_active_titles.different.sort_values(ascending=False)

title
Dirty Dancing (1987)                      0.830782
Good, The Bad and The Ugly, The (1966)    0.726351
Kentucky Fried Movie, The (1977)          0.676359
Jumpin' Jack Flash (1986)                 0.676359
Dumb & Dumber (1994)                      0.638608
                                            ...   
Robin Hood: Prince of Thieves (1991)      0.002056
Trainspotting (1996)                      0.001457
Jerry Maguire (1996)                      0.001109
Mystery, Alaska (1999)                    0.000997
Tender Mercies (1983)                     0.000142
Name: different, Length: 1216, dtype: float64

In [14]:
#   OR

mean_ratings_active_titles.sort_values(by="different" , ascending=False)

gender                                         F         M  different
title                                                                
Dirty Dancing (1987)                    3.790378  2.959596   0.830782
Good, The Bad and The Ugly, The (1966)  3.494949  4.221300   0.726351
Kentucky Fried Movie, The (1977)        2.878788  3.555147   0.676359
Jumpin' Jack Flash (1986)               3.254717  2.578358   0.676359
Dumb & Dumber (1994)                    2.697987  3.336595   0.638608
...                                          ...       ...        ...
Robin Hood: Prince of Thieves (1991)    3.161290  3.163347   0.002056
Trainspotting (1996)                    3.958974  3.960432   0.001457
Jerry Maguire (1996)                    3.758315  3.759424   0.001109
Mystery, Alaska (1999)                  3.434783  3.435780   0.000997
Tender Mercies (1983)                   3.905405  3.905263   0.000142

[1216 rows x 3 columns]

**++++ NOTE ++++**
<br> **"``.nlargest( )``" ::** Bu fonksiyon ile **;** bütün tabloyu sıralamak yerine **,** mesela **"** kadınlar ve erkekler arasındaki beğeni farkının EN FAZLA olduğu sadece **6** film **"** gibi bir bilgi isteniyorsa **,** bu fonksiyon daha pratik bir kullanım sağlıyor **:**
<br> Yani DataFrame'de bu fonksiyonun içerisine **,** hangi Kolondan kaç Satır görmek istediğimizi yazıyoruz. Bu fonksiyon da BÜYÜKTEN KÜÇÜĞE sıralayıp **,** yazdığımız Satır sayısı kadar Satırı bize gösteriyor... **:** 
<br> Nasıl kullanıldığına **, SHIFT + TAB** yaparak bakabiliriz...

In [17]:
mean_ratings_active_titles.nlargest(6 , "different")

gender                                         F         M  different
title                                                                
Dirty Dancing (1987)                    3.790378  2.959596   0.830782
Good, The Bad and The Ugly, The (1966)  3.494949  4.221300   0.726351
Kentucky Fried Movie, The (1977)        2.878788  3.555147   0.676359
Jumpin' Jack Flash (1986)               3.254717  2.578358   0.676359
Dumb & Dumber (1994)                    2.697987  3.336595   0.638608
Longest Day, The (1962)                 3.411765  4.031447   0.619682

**++++ NOTE ++++**
<br> **"``.nsmallest( )``" =** Bu fonksiyon ile **,** bütün tabloyu sıralamak yerine **,** mesela **"** kadınlar ve erkekler arasındaki beğeni farkının EN AZ olduğu sadece **5** film **"** gibi bir sonuç gerekiyorsa **,** bu fonksiyon daha pratik bir kullanım sağlıyor... **:**
<br> Yani DataFrame'de **,** bu fonksiyonun içerisine **,** hangi Kolondan kaç Satır görmek istediğimizi yazıyoruz. Bu fonksiyon da KÜÇÜKTEN BÜYÜĞE sıralayıp **,** yazdığımız Satır sayısı kadar Satırı bize gösteriyor... **:** 
<br> Nasıl kullanıldığına **, SHIFT + TAB** yaparak bakabiliriz...

In [18]:
mean_ratings_active_titles.nsmallest(5 , "different")

gender                                       F         M  different
title                                                              
Tender Mercies (1983)                 3.905405  3.905263   0.000142
Mystery, Alaska (1999)                3.434783  3.435780   0.000997
Jerry Maguire (1996)                  3.758315  3.759424   0.001109
Trainspotting (1996)                  3.958974  3.960432   0.001457
Robin Hood: Prince of Thieves (1991)  3.161290  3.163347   0.002056

In [48]:
mean_ratings_active_titles.head()

gender                                                     F         M  \
title                                                                    
American Beauty (1999)                              4.238901  4.347301   
Star Wars: Episode IV - A New Hope (1977)           4.302937  4.495307   
Star Wars: Episode V - The Empire Strikes Back ...  4.106481  4.344577   
Star Wars: Episode VI - Return of the Jedi (1983)   3.865237  4.069058   
Jurassic Park (1993)                                3.579407  3.814197   

gender                                              different  
title                                                          
American Beauty (1999)                               0.108400  
Star Wars: Episode IV - A New Hope (1977)            0.192371  
Star Wars: Episode V - The Empire Strikes Back ...   0.238096  
Star Wars: Episode VI - Return of the Jedi (1983)    0.203821  
Jurassic Park (1993)                                 0.234791

### **Most disagreement among viewers**

In [50]:
data.head()

user_id gender  age  occupation    zip  movie_id  rating  timestamp  \
0        1      F    1          10  48067      1193       5  978300760   
1        2      M   56          16  70072      1193       5  978298413   
2       12      M   25          12  32793      1193       4  978220179   
3       15      M   25           7  22903      1193       4  978199279   
4       17      M   50           1  95350      1193       5  978158471   

                                    title genres  
0  One Flew Over the Cuckoo's Nest (1975)  Drama  
1  One Flew Over the Cuckoo's Nest (1975)  Drama  
2  One Flew Over the Cuckoo's Nest (1975)  Drama  
3  One Flew Over the Cuckoo's Nest (1975)  Drama  
4  One Flew Over the Cuckoo's Nest (1975)  Drama

**++++ NOTE ++++**
<br> **``std( )`` ::** Bu fonksiyon ile **Standart Sapma** hesaplayacağız **:**

+ Calculate **Standart Deviation** for each movie **:**
<br> Her film için **Standart Sapmayı** hesapla **:**
<br> Yani her bir filmin **, ne kadar** farklı derecelerde oylandığını anlamaya çalışıyoruz.
<br> **+--- NOTE ---> 'Standart Sapma'** ne kadar **büyükse ,** o film o kadar **farklı** değerlerde reyting almıştır... **:**

In [20]:
ratings_std_by_title = data.groupby("title").rating.std() ;
ratings_std_by_title.head()

title
$1,000,000 Duck (1971)           1.092563
'Night Mother (1986)             1.118636
'Til There Was You (1997)        1.020159
'burbs, The (1989)               1.107760
...And Justice for All (1979)    0.878110
Name: rating, dtype: float64

In [14]:
#   OR

ratings_std_by_title = data.groupby("title")["rating"].std() ;
ratings_std_by_title.head()

title
$1,000,000 Duck (1971)           1.092563
'Night Mother (1986)             1.118636
'Til There Was You (1997)        1.020159
'burbs, The (1989)               1.107760
...And Justice for All (1979)    0.878110
Name: rating, dtype: float64

**++++ CRITICAL NOTE ++++**
<br> **.groupby( )** fonksiyonlarında fark ettiysek **,** parantez içerisine yazdığımız Kolon isimlerini hep **" "** tırnak içerisinde yazdık... ANCAK kolon ismi STRING değil de mesela INTEGER bir değer olsaydı o zaman tırnaksız olarak direkt sayının kendisini yazacaktık **!!!**

+ Select movies for only **'active_titles' :**
+ Sadece **"active_titles"** Kolonu için filmler seçin **:**
<br> Yani **, 250'den fazla** oy alan filmler için **'Standart Sapma'** işlemi yapalım **:**

In [21]:
ratings_std_by_active_titles = ratings_std_by_title.loc[active_titles]
ratings_std_by_active_titles

title
American Beauty (1999)                                            0.975133
Star Wars: Episode IV - A New Hope (1977)                         0.800781
Star Wars: Episode V - The Empire Strikes Back (1980)             0.844432
Star Wars: Episode VI - Return of the Jedi (1983)                 0.927356
Jurassic Park (1993)                                              0.969777
                                                                    ...   
Presidio, The (1988)                                              0.860205
Top Hat (1935)                                                    0.842724
Teenage Mutant Ninja Turtles II: The Secret of the Ooze (1991)    1.087971
U Turn (1997)                                                     1.146946
Random Hearts (1999)                                              1.035716
Name: rating, Length: 1216, dtype: float64

In [25]:
#   OR

#   SERİ üzerinde bir işlem yapmak istediğimiz için , 'LOC' fonksiyonunu kullanmasak da olurdu ::

ratings_std_by_active_titles[active_titles]

title
American Beauty (1999)                                            0.975133
Star Wars: Episode IV - A New Hope (1977)                         0.800781
Star Wars: Episode V - The Empire Strikes Back (1980)             0.844432
Star Wars: Episode VI - Return of the Jedi (1983)                 0.927356
Jurassic Park (1993)                                              0.969777
                                                                    ...   
Presidio, The (1988)                                              0.860205
Top Hat (1935)                                                    0.842724
Teenage Mutant Ninja Turtles II: The Secret of the Ooze (1991)    1.087971
U Turn (1997)                                                     1.146946
Random Hearts (1999)                                              1.035716
Name: rating, Length: 1216, dtype: float64

+ Şimdi de **,** bu **250'den fazla** reyting alan filmleri **, Standart Sapma**'larına göre büyükten küçüğe **sıralayalım :**
+ Ve sadece **ilk 10** filmi gösterelim **:**

In [26]:
ratings_std_by_active_titles.sort_values(ascending=False)[:10]

title
Dumb & Dumber (1994)                     1.321333
Blair Witch Project, The (1999)          1.316368
Natural Born Killers (1994)              1.307198
Tank Girl (1995)                         1.277695
Rocky Horror Picture Show, The (1975)    1.260177
Eyes Wide Shut (1999)                    1.259624
Evita (1996)                             1.253631
Billy Madison (1995)                     1.249970
Fear and Loathing in Las Vegas (1998)    1.246408
Bicentennial Man (1999)                  1.245533
Name: rating, dtype: float64

In [27]:
#   OR

#   "nlargerst()" fonksiyonu ile yapalım ::

ratings_std_by_active_titles.nlargest(10)

title
Dumb & Dumber (1994)                     1.321333
Blair Witch Project, The (1999)          1.316368
Natural Born Killers (1994)              1.307198
Tank Girl (1995)                         1.277695
Rocky Horror Picture Show, The (1975)    1.260177
Eyes Wide Shut (1999)                    1.259624
Evita (1996)                             1.253631
Billy Madison (1995)                     1.249970
Fear and Loathing in Las Vegas (1998)    1.246408
Bicentennial Man (1999)                  1.245533
Name: rating, dtype: float64

**++++ CRITICAL NOTE ++++**
<br> Büyük verilerle uğraşırken **, Sıralama** işleminde tüm veri setini sıralamak yerine **,** mesela İlk **10** veriye ihtiyacımız varsa **, sort_values( )** DEĞİL DE **nlargest( )** veya **nsmallest( )** kullanmak daha mantıklıdır... Bu sayede tüm veriyi çalıştırmak zorunda kalmayız **!** Ve program daha rahat çalışır **!**

### Create a **DataFrame** only from **'active_titles' :**
### (Yalnızca **'active_titles'**'dan bir **Veri Çerçevesi** oluşturun **:** )

+ Bütün tablodan **,** indirgenmiş bi tablo oluşturalım **,** ve filmler arasından **'active_titles'** filmlerini çekelim **:**
<br> Bunu **``.isin( )``** fonksiyonu ile yapacağız. Çünkü **'title' KOLONUNDA** bir **arama** yapmak istiyoruz... **:**

In [28]:
data.head()

user_id gender  age  occupation    zip  movie_id  rating  timestamp  \
0        1      F    1          10  48067      1193       5  978300760   
1        2      M   56          16  70072      1193       5  978298413   
2       12      M   25          12  32793      1193       4  978220179   
3       15      M   25           7  22903      1193       4  978199279   
4       17      M   50           1  95350      1193       5  978158471   

                                    title genres  
0  One Flew Over the Cuckoo's Nest (1975)  Drama  
1  One Flew Over the Cuckoo's Nest (1975)  Drama  
2  One Flew Over the Cuckoo's Nest (1975)  Drama  
3  One Flew Over the Cuckoo's Nest (1975)  Drama  
4  One Flew Over the Cuckoo's Nest (1975)  Drama

In [74]:
active_titles

Index(['American Beauty (1999)', 'Star Wars: Episode IV - A New Hope (1977)',
       'Star Wars: Episode V - The Empire Strikes Back (1980)',
       'Star Wars: Episode VI - Return of the Jedi (1983)',
       'Jurassic Park (1993)', 'Saving Private Ryan (1998)',
       'Terminator 2: Judgment Day (1991)', 'Matrix, The (1999)',
       'Back to the Future (1985)', 'Silence of the Lambs, The (1991)',
       ...
       'Bicycle Thief, The (Ladri di biciclette) (1948)',
       'Take the Money and Run (1969)', 'Bachelor, The (1999)',
       'Man Who Knew Too Little, The (1997)',
       'Wes Craven's New Nightmare (1994)', 'Presidio, The (1988)',
       'Top Hat (1935)',
       'Teenage Mutant Ninja Turtles II: The Secret of the Ooze (1991)',
       'U Turn (1997)', 'Random Hearts (1999)'],
      dtype='object', name='title', length=1216)

In [29]:
#   From data filter data for only 'active_titles' ::

data_reduced = data[data.title.isin(active_titles)] ;
data_reduced

user_id gender  age  occupation    zip  movie_id  rating  timestamp  \
0             1      F    1          10  48067      1193       5  978300760   
1             2      M   56          16  70072      1193       5  978298413   
2            12      M   25          12  32793      1193       4  978220179   
3            15      M   25           7  22903      1193       4  978199279   
4            17      M   50           1  95350      1193       5  978158471   
...         ...    ...  ...         ...    ...       ...     ...        ...   
983130     5816      M   25           0  70119      2889       5  957914832   
983131     5831      M   25           1  92120      2889       4  960873449   
983132     5836      M   25           0  91604      2889       3  957845106   
983133     5872      M   25          17  61265      2889       3  977282690   
983134     5880      M   25           1  08904      2889       3  962219824   

                                         title  genres  
0       One Flew Over the Cuckoo's Nest (1975)   Drama  
1       One Flew Over the Cuckoo's Nest (1975)   Drama  
2       One Flew Over the Cuckoo's Nest (1975)   Drama  
3       One Flew Over the Cuckoo's Nest (1975)   Drama  
4       One Flew Over the Cuckoo's Nest (1975)   Drama  
...                                        ...     ...  
983130                  Mystery, Alaska (1999)  Comedy  
983131                  Mystery, Alaska (1999)  Comedy  
983132                  Mystery, Alaska (1999)  Comedy  
983133                  Mystery, Alaska (1999)  Comedy  
983134                  Mystery, Alaska (1999)  Comedy  

[808922 rows x 10 columns]

+ Find how many titles you have in this "**data_reduced**" DataFrame **:**
<br> Yarattığımız bu "**data_reduced**" Veri Çerçevesinde kaç tane FARKLI film olduğunu bulun **:**

In [30]:
data_reduced.title.unique()

array(["One Flew Over the Cuckoo's Nest (1975)",
       'James and the Giant Peach (1996)', 'My Fair Lady (1964)', ...,
       'Last Temptation of Christ, The (1988)', 'Pet Sematary (1989)',
       'Mystery, Alaska (1999)'], dtype=object)

In [31]:
data_reduced.title.nunique()

1216

+ Do the same for Original data DataFrame **:**
<br> Aynısını **,** Orijinal Veri Çerçevesi için yapın **:**

In [84]:
data.title.nunique()

3706

+ Now find **top-rated 10 movies** using this **'data_reduced'** DataFrame **:**
<br> Şimdi bu **'data_reduced'** Veri Çerçevesini kullanarak **, en çok oy alan 10 filmi** bulun **:**

In [32]:
#   Calculate Average rating for each movie ::
#   Her film için Ortalama Reytingi hesaplayın ::
#   ("pivot_table" fonksiyonu , DEFAULT olarak Ortalama hesaplar!)

top_rated = data_reduced.pivot_table("rating" , index="title").nlargest(10 , "rating")
top_rated

rating
title                                                       
Seven Samurai (The Magnificent Seven) (Shichini...  4.560510
Shawshank Redemption, The (1994)                    4.554558
Godfather, The (1972)                               4.524966
Close Shave, A (1995)                               4.520548
Usual Suspects, The (1995)                          4.517106
Schindler's List (1993)                             4.510417
Wrong Trousers, The (1993)                          4.507937
Sunset Blvd. (a.k.a. Sunset Boulevard) (1950)       4.491489
Raiders of the Lost Ark (1981)                      4.477725
Rear Window (1954)                                  4.476190

In [105]:
#   OR

top_rated = data_reduced.pivot_table("rating" , index="title").sort_values(by="rating" , ascending=False).head(10)
top_rated

rating
title                                                       
Seven Samurai (The Magnificent Seven) (Shichini...  4.560510
Shawshank Redemption, The (1994)                    4.554558
Godfather, The (1972)                               4.524966
Close Shave, A (1995)                               4.520548
Usual Suspects, The (1995)                          4.517106
Schindler's List (1993)                             4.510417
Wrong Trousers, The (1993)                          4.507937
Sunset Blvd. (a.k.a. Sunset Boulevard) (1950)       4.491489
Raiders of the Lost Ark (1981)                      4.477725
Rear Window (1954)                                  4.476190

**++++ CRITICAL NOTE ++++**
<br> Yukarıda fark ettiysek **, (index="title")** şeklinde yazdık... **pivot_table( )** fonksiyonunun içerisinde SHIFT + TAB yaptığımızda **,** o açıklamadaki **sırayla** hepsini tek tek yazarsak **,** mesela **(index=...)** diye belirtmemize gerek kalmıyor... ANCAK sırayı bozarak yazmak istersek veya aralarından bazılarını atlayıp yazmak istersek **,** o zaman belirtmek ZORUNDAYIZ **!**

+ Now do the above by finding also the number of rating for each movie **:**
<br> Şimdi her film için reyting sayısını da bularak yukarıdakileri yapın **:**
<br> Yani her bir filmin **Ortalama reytingini** hesapladık ama acaba bu filmlerin her biri toplam kaç kişi tarafından oylanmış **?** Yani her bir filmde **kaç tane reyting** verisi var **?**

In [62]:
#   Now we want to both calculate average rating and the number of rating for each movie ::

#   Şimdi , KÖŞELİ Parantez İle (aggfunc=['mean', 'count']) ; her film için hem Ortalama reytingi hem de
# reyting Sayısını hesaplamak istiyoruz ::

top_rated = data_reduced.pivot_table("rating" , "title" , aggfunc=['mean', 'count']) ;
top_rated

mean  count
                                     rating rating
title                                             
'burbs, The (1989)                 2.910891    303
10 Things I Hate About You (1999)  3.422857    700
101 Dalmatians (1961)              3.596460    565
101 Dalmatians (1996)              3.046703    364
12 Angry Men (1957)                4.295455    616
...                                     ...    ...
Young Guns (1988)                  3.418149    562
Young Guns II (1990)               2.907859    369
Young Sherlock Holmes (1985)       3.390501    379
Zero Effect (1998)                 3.750831    301
eXistenZ (1999)                    3.256098    410

[1216 rows x 2 columns]

+ Now sort values by "**count**" column **:**
<br> (Note that resulting table has two levels in columns.)
<br> **+--- NOTE --->** When you want to sort you should provide in **tuple** format...
+ Şimdi değerleri **"count" (sayı)** Sütununa göre sıralayın **:**
<br> (Yukarıdaki tablonun Sütunlarının **,** iki düzeyi olduğunu yani **multi-index** olduğunu unutmayın.)
<br> **+--- NOTE ---> Sıralama** işlemi yapmak istiyorsak **; multi-index** Kolonun ismini **; tuple** biçiminde yazmamız gerekiyor...

In [55]:
top_rated.nlargest(10 , ("count", "rating"))

mean  count
                                                      rating rating
title                                                              
American Beauty (1999)                              4.317386   3428
Star Wars: Episode IV - A New Hope (1977)           4.453694   2991
Star Wars: Episode V - The Empire Strikes Back ...  4.292977   2990
Star Wars: Episode VI - Return of the Jedi (1983)   4.022893   2883
Jurassic Park (1993)                                3.763847   2672
Saving Private Ryan (1998)                          4.337354   2653
Terminator 2: Judgment Day (1991)                   4.058513   2649
Matrix, The (1999)                                  4.315830   2590
Back to the Future (1985)                           3.990321   2583
Silence of the Lambs, The (1991)                    4.351823   2578

**++++ NOTE ++++**
+ Yukarıdaki işlemi ayrı ayrı uğraşmadan tek kod satırında da bitirebilirdik **:**

In [56]:
top_rated = data_reduced.pivot_table("rating" , "title" , aggfunc=["mean", "count"]).nlargest(10 , ("count", "rating"))
top_rated

mean  count
                                                      rating rating
title                                                              
American Beauty (1999)                              4.317386   3428
Star Wars: Episode IV - A New Hope (1977)           4.453694   2991
Star Wars: Episode V - The Empire Strikes Back ...  4.292977   2990
Star Wars: Episode VI - Return of the Jedi (1983)   4.022893   2883
Jurassic Park (1993)                                3.763847   2672
Saving Private Ryan (1998)                          4.337354   2653
Terminator 2: Judgment Day (1991)                   4.058513   2649
Matrix, The (1999)                                  4.315830   2590
Back to the Future (1985)                           3.990321   2583
Silence of the Lambs, The (1991)                    4.351823   2578

In [53]:
#   OR

#   SÜSLÜ Parantez İle (aggfunc={'mean', 'count'}) , her film için hem Ortalama reytingi hem de reyting Sayısını
# hesaplamak istiyoruz ::
#   Ve sadece , ORTALAMASI ('mean') en yüksek olan İlk 10 filmi görmek istiyoruz ::

top_rated = data_reduced.pivot_table("rating" , "title" , aggfunc={'mean', 'count'}).nlargest(10 , "mean") ;
top_rated

count      mean
title                                                               
Seven Samurai (The Magnificent Seven) (Shichini...   628.0  4.560510
Shawshank Redemption, The (1994)                    2227.0  4.554558
Godfather, The (1972)                               2223.0  4.524966
Close Shave, A (1995)                                657.0  4.520548
Usual Suspects, The (1995)                          1783.0  4.517106
Schindler's List (1993)                             2304.0  4.510417
Wrong Trousers, The (1993)                           882.0  4.507937
Sunset Blvd. (a.k.a. Sunset Boulevard) (1950)        470.0  4.491489
Raiders of the Lost Ark (1981)                      2514.0  4.477725
Rear Window (1954)                                  1050.0  4.476190

**++++ NOTE ++++**
<br> Yukarıdaki işlemi **, Süslü Parantez ile** yaptığımız için **,**  **tuple** kullanmadan yapabildik... Çünkü yukarıdaki **,** son sonuç DataFrame'inde Kolon ismi olarak SADECE **"count"** veya **"mean"** yazıyor... Böyle olunca da sorunsuz şekilde bu işlemi bitirebiliyoruz...
<br> Aynı işlemi **Köşeli Parantez ile** yaptığımızda **, tuple** kullanmamız gerekiyor... Çünkü **, Köşeli Parantez** ile yazdığımızda OUT olarak DataFrame'de Kolon isimleri **, "mean / rating"** veya **"count / rating"** şeklinde yazıyor... Yani kolonlar **, MULTI-INDEX** olarak görünüyor **!** Yani **Multi-Index** Tablolarda bir Kolonu ifade edebilmek için **tuple** içerisinde yazmamız gerekiyor...

**OR**
+ Change the Columns of the "**top_rated**" DataFrame **:**
<br> Keep only **mean** and **count** Columns **:**
+ "**top_rated**" Veri Çerçevesinin Sütunlarını değiştirin **:**
<br> Yani Kolonları **("main / rating")** halinde **,** yani **multi-index** şeklinde DEĞİL DE **; "rating"** olmadan **,** yalnızca **"mean"** ve **"count"** halinde **,** tek index olacak şekilde değiştir **:**

In [45]:
top_rated.head()

count      mean
title                                                               
Seven Samurai (The Magnificent Seven) (Shichini...   628.0  4.560510
Shawshank Redemption, The (1994)                    2227.0  4.554558
Godfather, The (1972)                               2223.0  4.524966
Close Shave, A (1995)                                657.0  4.520548
Usual Suspects, The (1995)                          1783.0  4.517106

In [51]:
top_rated.columns=["mean" , "count"]
top_rated.head()

mean     count
title                                                               
Seven Samurai (The Magnificent Seven) (Shichini...   628.0  4.560510
Shawshank Redemption, The (1994)                    2227.0  4.554558
Godfather, The (1972)                               2223.0  4.524966
Close Shave, A (1995)                                657.0  4.520548
Usual Suspects, The (1995)                          1783.0  4.517106

**OR**
+ **``.columns.drop_duplicates( )`` =** Bu komut ile **,** Bir DataFrame'de **aynı isimde** Kolonlar varsa **,** onları **düşürebiliyoruz (silebiliyoruz)...**
+ **``.columns.droplevel( )`` =** Bu komut ile **,** SERİ'de veya DataFrame'de **multi-index** olan Kolonlarda **,** KAÇINCI SEVİYE'de olan Kolonu **silmek istiyorsak , silebiliyoruz (düşürebiliyoruz) :**

In [73]:
top_rated = data_reduced.pivot_table("rating" , "title" , aggfunc=["mean", "count"]).nlargest(10 , ("count", "rating"))
top_rated.head(2)

mean  count
                                             rating rating
title                                                     
American Beauty (1999)                     4.317386   3428
Star Wars: Episode IV - A New Hope (1977)  4.453694   2991

In [74]:
top_rated.columns

MultiIndex([( 'mean', 'rating'),
            ('count', 'rating')],
           )

In [75]:
top_rated.columns = top_rated.columns.droplevel(1);
top_rated.head(2)

mean  count
title                                                     
American Beauty (1999)                     4.317386   3428
Star Wars: Episode IV - A New Hope (1977)  4.453694   2991

In [76]:
top_rated.columns

Index(['mean', 'count'], dtype='object')

### Create a **pivot table** for average movie rating for each occupation **:**
### (Her mesleğin ortalama film reytingi için bir **pivot_table** oluşturun **:** )

In [46]:
data_reduced.head()

user_id gender  age  occupation    zip  movie_id  rating  timestamp  \
0        1      F    1          10  48067      1193       5  978300760   
1        2      M   56          16  70072      1193       5  978298413   
2       12      M   25          12  32793      1193       4  978220179   
3       15      M   25           7  22903      1193       4  978199279   
4       17      M   50           1  95350      1193       5  978158471   

                                    title genres  
0  One Flew Over the Cuckoo's Nest (1975)  Drama  
1  One Flew Over the Cuckoo's Nest (1975)  Drama  
2  One Flew Over the Cuckoo's Nest (1975)  Drama  
3  One Flew Over the Cuckoo's Nest (1975)  Drama  
4  One Flew Over the Cuckoo's Nest (1975)  Drama

In [77]:
#   Each line shows a movie and each column shows an occupation ::
#   Aşağıda , Her Satırda bir film ve Her Sütunda bir meslek gösteriliyor ::

avg_rating_by_occupation = data_reduced.pivot_table("rating" , "title" , "occupation")

In [78]:
avg_rating_by_occupation.head()

occupation                               0         1         2         3   \
title                                                                       
'burbs, The (1989)                 2.580645  2.958333  2.950000  2.833333   
10 Things I Hate About You (1999)  3.337079  3.463415  3.285714  3.478261   
101 Dalmatians (1961)              3.546667  3.627451  3.625000  3.736842   
101 Dalmatians (1996)              3.050847  3.117647  2.944444  3.166667   
12 Angry Men (1957)                4.318841  4.148148  4.333333  4.400000   

occupation                               4         5         6         7   \
title                                                                       
'burbs, The (1989)                 3.022222  2.916667  2.625000  2.615385   
10 Things I Hate About You (1999)  3.391026  3.769231  3.428571  3.558140   
101 Dalmatians (1961)              3.423529  3.666667  3.736842  3.716981   
101 Dalmatians (1996)              2.595238  3.333333  3.437500  3.190476   
12 Angry Men (1957)                4.115385  4.400000  4.516129  4.230769   

occupation                          8         9   ...        11        12  \
title                                             ...                       
'burbs, The (1989)                 5.0  2.800000  ...  2.000000  2.933333   
10 Things I Hate About You (1999)  4.0  3.916667  ...  3.200000  3.250000   
101 Dalmatians (1961)              4.0  3.916667  ...  4.444444  3.454545   
101 Dalmatians (1996)              3.0  2.909091  ...  3.200000  2.916667   
12 Angry Men (1957)                5.0  4.400000  ...  4.409091  4.380952   

occupation                               13        14        15        16  \
title                                                                       
'burbs, The (1989)                      NaN  3.250000  3.200000  2.833333   
10 Things I Hate About You (1999)  4.000000  3.388889  3.076923  3.375000   
101 Dalmatians (1961)              3.142857  3.478261  4.111111  3.521739   
101 Dalmatians (1996)              3.666667  3.071429  4.000000  3.375000   
12 Angry Men (1957)                4.368421  4.333333  4.571429  4.366667   

occupation                               17        18        19        20  
title                                                                      
'burbs, The (1989)                 3.000000  3.100000  3.400000  3.000000  
10 Things I Hate About You (1999)  3.510638  3.571429  3.090909  3.052632  
101 Dalmatians (1961)              3.641026  3.625000  2.857143  3.709677  
101 Dalmatians (1996)              3.352941  3.400000  3.200000  2.571429  
12 Angry Men (1957)                4.333333  4.625000  4.428571  4.230769  

[5 rows x 21 columns]

In [79]:
#   OR

#   Burada da , ORTALAMA HESAPLAMA fonksiyonunu hazır kullanmayalım da "lambda" fonksiyonu ile kendimiz yazalım ::

avg_rating_by_occupation=data_reduced.pivot_table("rating", "title", "occupation", aggfunc=lambda x: sum(x)/len(x))

In [43]:
avg_rating_by_occupation.head()

occupation                               0         1         2         3   \
title                                                                       
'burbs, The (1989)                 2.580645  2.958333  2.950000  2.833333   
10 Things I Hate About You (1999)  3.337079  3.463415  3.285714  3.478261   
101 Dalmatians (1961)              3.546667  3.627451  3.625000  3.736842   
101 Dalmatians (1996)              3.050847  3.117647  2.944444  3.166667   
12 Angry Men (1957)                4.318841  4.148148  4.333333  4.400000   

occupation                               4         5         6         7   \
title                                                                       
'burbs, The (1989)                 3.022222  2.916667  2.625000  2.615385   
10 Things I Hate About You (1999)  3.391026  3.769231  3.428571  3.558140   
101 Dalmatians (1961)              3.423529  3.666667  3.736842  3.716981   
101 Dalmatians (1996)              2.595238  3.333333  3.437500  3.190476   
12 Angry Men (1957)                4.115385  4.400000  4.516129  4.230769   

occupation                          8         9   ...        11        12  \
title                                             ...                       
'burbs, The (1989)                 5.0  2.800000  ...  2.000000  2.933333   
10 Things I Hate About You (1999)  4.0  3.916667  ...  3.200000  3.250000   
101 Dalmatians (1961)              4.0  3.916667  ...  4.444444  3.454545   
101 Dalmatians (1996)              3.0  2.909091  ...  3.200000  2.916667   
12 Angry Men (1957)                5.0  4.400000  ...  4.409091  4.380952   

occupation                               13        14        15        16  \
title                                                                       
'burbs, The (1989)                      NaN  3.250000  3.200000  2.833333   
10 Things I Hate About You (1999)  4.000000  3.388889  3.076923  3.375000   
101 Dalmatians (1961)              3.142857  3.478261  4.111111  3.521739   
101 Dalmatians (1996)              3.666667  3.071429  4.000000  3.375000   
12 Angry Men (1957)                4.368421  4.333333  4.571429  4.366667   

occupation                               17        18        19        20  
title                                                                      
'burbs, The (1989)                 3.000000  3.100000  3.400000  3.000000  
10 Things I Hate About You (1999)  3.510638  3.571429  3.090909  3.052632  
101 Dalmatians (1961)              3.641026  3.625000  2.857143  3.709677  
101 Dalmatians (1996)              3.352941  3.400000  3.200000  2.571429  
12 Angry Men (1957)                4.333333  4.625000  4.428571  4.230769  

[5 rows x 21 columns]

### Find the top-10 movie that has highest **Standart Deviation** by different occupations **:**
### (Farklı mesleklere göre en yüksek **Standart Sapmaya** sahip ilk 10 filmi bulun **:** )
+ **Yani hangi filmin reytingleri , mesleklere göre daha çok değişiyor bunu bulalım :**

In [80]:
std_by_occupation = avg_rating_by_occupation.std(axis=1)

**+--- NOTE --->** Yukarıda **, axis=1** yazmamızın sebebi **;** DEFAULT halde **"axis=0"** olduğunda **,** her bir KOLONUN **,** yani mesleğin kendi içinde her bir elemanı için bir hesaplama yapıyor... ANCAK biz her bir filmin **,** yani her bir SATIRIN kendi içinde her bir elemanı için bir hesaplama yapmak istiyoruz...

In [81]:
std_by_occupation.head()

title
'burbs, The (1989)                   0.712480
10 Things I Hate About You (1999)    0.291116
101 Dalmatians (1961)                0.325618
101 Dalmatians (1996)                0.328754
12 Angry Men (1957)                  0.209509
dtype: float64

+ Now sort this series and find the **top-10** movies **:**
<br> Şimdi bu diziyi Sıralayın ve **en iyi 10** filmi (Yani **Standart Sapma**'sı EN AZ olan **10** filmi) bulun **:**

In [47]:
#   Display top-10 movies having the lowest Standart Deviation ::
#   En Düşük Standart Sapmaya sahip ilk 10 filmi görüntüleyin ::

std_by_occupation.sort_values().head(10)

title
Fargo (1996)                        0.092501
Raiders of the Lost Ark (1981)      0.106728
Shawshank Redemption, The (1994)    0.109251
Broadcast News (1987)               0.113001
Superman (1978)                     0.115835
Ghostbusters (1984)                 0.116602
Shakespeare in Love (1998)          0.117459
Schindler's List (1993)             0.117710
Men in Black (1997)                 0.117894
Goldfinger (1964)                   0.118013
dtype: float64

+ Şimdi de **Standart Sapması** EN FAZLA olan **5** filmi bulalım **:**

In [75]:
#   Display top-10 movies having the highest Standart Deviation ::
#   En Yüksek Standart Sapmaya sahip ilk 10 filmi görüntüleyin ::

std_by_occupation.sort_values(ascending=False).head()

title
Tank Girl (1995)                           0.803774
Immortal Beloved (1994)                    0.802270
Some Folks Call It a Sling Blade (1993)    0.799291
Super Mario Bros. (1993)                   0.779125
Top Hat (1935)                             0.778740
dtype: float64

In [79]:
# OR

std_by_occupation.nlargest(5)

title
Tank Girl (1995)                           0.803774
Immortal Beloved (1994)                    0.802270
Some Folks Call It a Sling Blade (1993)    0.799291
Super Mario Bros. (1993)                   0.779125
Top Hat (1935)                             0.778740
dtype: float64

## Extract the **Year** information from the title **:**
## (Başlıktan **Yıl** bilgisini çıkarın **:** )

+ An easy way **::** Get the title column and take the year indexing from the right side **:**
<br> (You can use **lambda** function and **map** fuctions **,** )
<br> (Create a new column with a name **"year" ,** )
<br> (Work on **"data_reduced"** DataFrame **.** )
+ Kolay bir yol **::** Başlık Sütununu alın ve sağ taraftan da Yıl indekslemeyi alın **:**
<br> (**lambda** fonksiyonunu ve **map** fonksiyonunu kullanabilirsiniz **,** )
<br> (**"year"** adında yeni bir sütun oluşturun **,** )
<br> (**"data_reduced"** Veri Çerçevesi üzerinde çalışın **.** )

In [49]:
data_reduced.head()

user_id gender  age  occupation    zip  movie_id  rating  timestamp  \
0        1      F    1          10  48067      1193       5  978300760   
1        2      M   56          16  70072      1193       5  978298413   
2       12      M   25          12  32793      1193       4  978220179   
3       15      M   25           7  22903      1193       4  978199279   
4       17      M   50           1  95350      1193       5  978158471   

                                    title genres  
0  One Flew Over the Cuckoo's Nest (1975)  Drama  
1  One Flew Over the Cuckoo's Nest (1975)  Drama  
2  One Flew Over the Cuckoo's Nest (1975)  Drama  
3  One Flew Over the Cuckoo's Nest (1975)  Drama  
4  One Flew Over the Cuckoo's Nest (1975)  Drama

+ Mesela **,** yukarıda gördüğümüz gibi **, One Flew Over the Cuckoo's Nest (1975)** filminin sonunda **,** yapıldığı yıl yazıyor... Bu şekilde **"title"** kolonunda yazan film isimlerinin içerisindeki Yıl bölümünü alıp **, "year"** diye **yeni** bir Kolon açıp **,** oraya yerleştirelim **:**
+ + Bunu da mesela burada nasıl yapabiliriz **?** Veri Setine baktığımızda **"title"** Kolonundaki verilerin **son 6 String** ifadesinin **, Parantez** ve **Yıl** verisi içerdiğini görüyoruz... Buradan bir sonuç çıkartıp **genel** bir işlem yapabiliriz... **:**

---
---
**---------->**

In [84]:
#   ÖRNEK :

text = "One Flew Over the Cuckoo's Nest (1975)"
text[ -5 : -1 ]

'1975'

In [85]:
#   OR :

text[ -6: ]

'(1975)'

**<----------**

---
---

**++++ NOTE ++++**
<br> **``str.slice( )`` ::** Bu metod **;** yaptığımız işlemi **,** BÜTÜN KOLONA yaymamızı sağlıyor... Yani içerisine yazdığımız işlemi **,** Default olarak KOLONUN TÜM ELEMANLARINA uyguluyor **:**

In [86]:
data_reduced["year"] = data_reduced.title.str.slice(-5 , -1)

<ipython-input-86-a4c9c8624c03>:1: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  data_reduced["year"] = data_reduced.title.str.slice(-5 , -1)


In [87]:
data_reduced.head()

user_id gender  age  occupation    zip  movie_id  rating  timestamp  \
0        1      F    1          10  48067      1193       5  978300760   
1        2      M   56          16  70072      1193       5  978298413   
2       12      M   25          12  32793      1193       4  978220179   
3       15      M   25           7  22903      1193       4  978199279   
4       17      M   50           1  95350      1193       5  978158471   

                                    title genres  year  
0  One Flew Over the Cuckoo's Nest (1975)  Drama  1975  
1  One Flew Over the Cuckoo's Nest (1975)  Drama  1975  
2  One Flew Over the Cuckoo's Nest (1975)  Drama  1975  
3  One Flew Over the Cuckoo's Nest (1975)  Drama  1975  
4  One Flew Over the Cuckoo's Nest (1975)  Drama  1975

**+--- CRITICAL NOTE --->** Yukarıda fark ettiysek, **(-5 , -1)** şeklinde **,** yani aralarına iki nokta üst üste koyarak değil de **,** VİRGÜL koyarak yaptık...

**+--- CRITICAL NOTE --->** Buradaki "**.str.slice( )**" metodundaki problem de **,** eğer Veri Setinde herhangi bir problem varsa **,** mesela verilerde hemen gözle görünemeyen bir boşluk vb. gereksiz bir karakter varsa **,** istemediğimiz bir sonuç alabiliriz... Veya filmlerin içerisinden **,** ismi direkt sadece bir Tarih içeren filmler varsa **,** orada da problem yaşayabiliriz...

#### Bu işlemi **;** başka bir yol olan **,** herhangi bir **"Regular Expression"** metodu ile gerçekleştirelim **:**
<br> (**"``.str``"** fonksiyonu **;** tamamen kendine özgü metodları olduğu gibi **Regular Expression** metodları da içerir)
<br> Mesela bunlardan biri de **, "``.str.findall( )``"** metodu **:**
<br> **[ ``.findall( )``** metodu da **,** bir **``pattern``** belirlenip **,** içerisine yazılarak kullanılıyor... **]**

**+--- NOTE --->** Normalde **parantez** karakterinin **,** STRING ifadelerde PATTERN hazırlarken özel bir anlamı (kullanımı) var... Ama biz direkt **parantez** karakterinin kendisini kullanmak istiyorsak **,** kural içerisinde önüne **" ``\`` "** karakteri koyarak o şekilde yazmalıyız... Yani **, "``\ (``"** şeklinde yazmalıyız...
<br> **+--- NOTE --->** Mesela **,** Parantez sonrasında rakam da olsun istiyorsak... Bunun için de **, "``\d``"** karakterini kullanıyoruz... Rakam sayısının da TAMI TAMINA **4** tane olmasını istiyorsak **"``\d``"** sonrasında **, " ``{4}`` "** yazmamız gerekiyor... Yani sonuçta TAM OLARAK **4** tane rakam olsun istiyorsak **, "``\d{4}``"** şeklinde yazmamız gerekiyor...

In [94]:
data_reduced["year"] = data_reduced.title.str.findall("\(\d{4}\)")

<ipython-input-94-f4494144f659>:1: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  data_reduced["year"] = data_reduced.title.str.findall("\(\d{4}\)")


In [95]:
data_reduced.head()

user_id gender  age  occupation    zip  movie_id  rating  timestamp  \
0        1      F    1          10  48067      1193       5  978300760   
1        2      M   56          16  70072      1193       5  978298413   
2       12      M   25          12  32793      1193       4  978220179   
3       15      M   25           7  22903      1193       4  978199279   
4       17      M   50           1  95350      1193       5  978158471   

                                    title genres      year  
0  One Flew Over the Cuckoo's Nest (1975)  Drama  [(1975)]  
1  One Flew Over the Cuckoo's Nest (1975)  Drama  [(1975)]  
2  One Flew Over the Cuckoo's Nest (1975)  Drama  [(1975)]  
3  One Flew Over the Cuckoo's Nest (1975)  Drama  [(1975)]  
4  One Flew Over the Cuckoo's Nest (1975)  Drama  [(1975)]

**++++ NOTE ++++**
<br> **Peki , PATTERN'de PARANTEZ ifadesinin GÖREVİ nedir ?**
<br> Parantez ifadesi **;** MESELA **,** AD **=** Alper **-** SOYAD **=** Ceylan ifadesi gibi bir **eşleştirme** yapmak istiyoruz... Bunun için **, " AD=\w+ "** şeklinde bir PATTERN oluşturmamız gerekiyor... Burada **" AD="** kısmının görünmeyip **,** sadece ismin kendisinin görünmesini istiyorsak **,** bu kuralı **" ( ) "** içerisinde yazmamız gerekiyor... Yani **,** SADECE görünmesini istediğimiz yeri **,** PATTERN'de PARANTEZ içerisine almamız gerekiyor. Yani **, "``AD=(\w+)``"** şeklinde yazmamız gerekiyor... **:**

In [33]:
data_reduced["year"] = data_reduced.title.str.findall("\((\d{4})\)");

<ipython-input-33-667f5bc3b537>:1: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  data_reduced["year"] = data_reduced.title.str.findall("\((\d{4})\)");


In [34]:
data_reduced.head()

user_id gender  age  occupation    zip  movie_id  rating  timestamp  \
0        1      F    1          10  48067      1193       5  978300760   
1        2      M   56          16  70072      1193       5  978298413   
2       12      M   25          12  32793      1193       4  978220179   
3       15      M   25           7  22903      1193       4  978199279   
4       17      M   50           1  95350      1193       5  978158471   

                                    title genres    year  
0  One Flew Over the Cuckoo's Nest (1975)  Drama  [1975]  
1  One Flew Over the Cuckoo's Nest (1975)  Drama  [1975]  
2  One Flew Over the Cuckoo's Nest (1975)  Drama  [1975]  
3  One Flew Over the Cuckoo's Nest (1975)  Drama  [1975]  
4  One Flew Over the Cuckoo's Nest (1975)  Drama  [1975]

Yukarıda gördüğümüz gibi **,** artık Yıl değeri parantez içerisinde değil... Parantezleri PATTERN'de belirlediğimiz üzere dışarıda bıraktık... Sadece Köşeli Parantez kaldı...

**+--- NOTE --->** OUT olarak **"year"** Kolonunda verilerin Köşeli Parantez içerisinde yazmasının sebebi de **,** "**.findall( )**" metodunun OUT olarak LİSTE döndürmesidir...

+ KÖŞELİ PARANTEZ'den de kurtulmak istiyorsak **,** yani LİSTE içerisine girmeden bu işlemi gerçekleştirmesini istiyorsak **,** yine STRING fonksiyonunun metodlarından birini kullanabiliriz **:**
<br> **( ``.str.extract( )``** metodu **)**

In [96]:
data_reduced["year"] = data_reduced.title.str.extract("\((\d{4})\)")

<ipython-input-96-d902668ffc6d>:1: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  data_reduced["year"] = data_reduced.title.str.extract("\((\d{4})\)")


In [97]:
data_reduced.head()

user_id gender  age  occupation    zip  movie_id  rating  timestamp  \
0        1      F    1          10  48067      1193       5  978300760   
1        2      M   56          16  70072      1193       5  978298413   
2       12      M   25          12  32793      1193       4  978220179   
3       15      M   25           7  22903      1193       4  978199279   
4       17      M   50           1  95350      1193       5  978158471   

                                    title genres  year  
0  One Flew Over the Cuckoo's Nest (1975)  Drama  1975  
1  One Flew Over the Cuckoo's Nest (1975)  Drama  1975  
2  One Flew Over the Cuckoo's Nest (1975)  Drama  1975  
3  One Flew Over the Cuckoo's Nest (1975)  Drama  1975  
4  One Flew Over the Cuckoo's Nest (1975)  Drama  1975

---
---
**---------->**

+ Yukarıda bahsettiğim **' .str.slice( ) '** metodunun kullanılmasında yaşanabilecek sorunların var olup olmadığı ile ilgili bir araştırma yapalım... Mesela kendini tekrar eden yani birden fazla olan veriler var mı **?** Veya içerisinde Tarih gibi **,** yan yana **4** tane rakamın olduğu bir film var mı **?** Bakalım **:**

In [98]:
result = data_reduced.title.str.findall(r"\(\d{4}\)").map(lambda x: len(x))
result

0         1
1         1
2         1
3         1
4         1
         ..
983130    1
983131    1
983132    1
983133    1
983134    1
Name: title, Length: 808922, dtype: int64

In [38]:
result[result != 1]

Series([], Name: title, dtype: int64)

Yukarıda gördüğümüz gibi **1** dışında bir eleman bulamadı... Yani kendini tekrar eden bir eleman yokmuş...

In [39]:
result2 = data_reduced.title.str.findall(r"\d{4}").map(lambda x: len(x))
result2[result2 != 1]

446247    2
446248    2
446249    2
446250    2
446251    2
         ..
813845    2
813846    2
813847    2
813848    2
813849    2
Name: title, Length: 3376, dtype: int64

In [40]:
data_reduced[result2 != 1].head()

user_id gender  age  occupation    zip  movie_id  rating  timestamp  \
446247       10      F   35           1  95370       671       3  978228408   
446248       22      M   18          15  53706       671       5  978134995   
446249       23      M   35           0  90049       671       3  978464728   
446250       41      F   18           4  15116       671       5  978039548   
446251       45      F   45          16  94110       671       3  977990487   

                                                 title         genres  year  
446247  Mystery Science Theater 3000: The Movie (1996)  Comedy|Sci-Fi  1996  
446248  Mystery Science Theater 3000: The Movie (1996)  Comedy|Sci-Fi  1996  
446249  Mystery Science Theater 3000: The Movie (1996)  Comedy|Sci-Fi  1996  
446250  Mystery Science Theater 3000: The Movie (1996)  Comedy|Sci-Fi  1996  
446251  Mystery Science Theater 3000: The Movie (1996)  Comedy|Sci-Fi  1996

Yukarıda gördüğümüz gibi de **,** film adı olarak **, 4** tane rakam içeren filmler varmış...
+ Şimdi de bu filmleri bulup ortaya çıkaralım **:**

In [41]:
data_reduced[result2 != 1].drop_duplicates("title")

user_id gender  age  occupation    zip  movie_id  rating  timestamp  \
446247       10      F   35           1  95370       671       3  978228408   
483870       10      F   35           1  95370       924       3  979168183   
598171       15      M   25           7  22903      1422       4  978212463   
632788       17      M   50           1  95350      2311       4  978161068   
813397       32      F   25           0  19355      3159       2  978122583   

                                                 title  \
446247  Mystery Science Theater 3000: The Movie (1996)   
483870                    2001: A Space Odyssey (1968)   
598171                           Murder at 1600 (1997)   
632788                                     2010 (1984)   
813397                            Fantasia 2000 (1999)   

                               genres  year  
446247                  Comedy|Sci-Fi  1996  
483870  Drama|Mystery|Sci-Fi|Thriller  1968  
598171               Mystery|Thriller  1997  
632788                 Mystery|Sci-Fi  1984  
813397   Animation|Children's|Musical  1999

Yani yukarıda gördüğümüz gibi **,** bize problem çıkarabilecek olan filmler yukarıdakilermiş...

<----------

---
---

### Find the average rating by movie year **:**
### (Film yılına göre ortalama reytingi bulun **:** )
+ Yani her Yılda **,** filmler Ortalama Kaç reyting almışlar **?** Bunu bulalım **:**

In [99]:
avg_rating_by_year = data_reduced.pivot_table("rating" , "year")

In [100]:
avg_rating_by_year.head()

rating
year          
1925  4.189091
1926  4.082474
1930  4.194030
1931  4.190863
1933  3.900000

+ Sort this table and find the top-10 years having the highest average rating score **:**
<br> Bu tabloyu **sıralayın** ve **en yüksek Ortalama reyting** puanına sahip **ilk 5 Yılı** bulun **:**

In [101]:
avg_rating_by_year.nlargest(5 , "rating")

rating
year          
1949  4.452083
1946  4.302041
1944  4.264686
1934  4.262763
1952  4.246967

### Find top-rated movies at each year **:**
### (Her yıl en çok reyting (oy) (puan) alan filmleri bulun **:** )
+ Yani mesela **1975** yılında yapılan filmler arasında **en çok reyting alan** filmler hangileri **?**

+ First find average rating by **"title" :**
<br> İlk önce **, "title"** (başlık) Kolonuna göre **Ortalama reytingi** bulalım **:**

In [102]:
avg_rating_by_title = data_reduced.pivot_table("rating" , "title");
avg_rating_by_title.head()

rating
title                                      
'burbs, The (1989)                 2.910891
10 Things I Hate About You (1999)  3.422857
101 Dalmatians (1961)              3.596460
101 Dalmatians (1996)              3.046703
12 Angry Men (1957)                4.295455

+ Mesela yukarıdaki DataFrame'deki filmlerin Tarihini çekmek istiyoruz...
<br> Bunun çok fazla yolu var... Biz **, index**'teki verilerin üzerinde STRING fonksiyonunun bir çok metodunu uygulayabiliriz... **:**
<br> **(Bu çok güzel bir özelliktir...)**

In [108]:
avg_rating_by_title["year"] = avg_rating_by_title.index.str.slice(-5 , -1)

In [109]:
avg_rating_by_title.head()

rating  year
title                                            
'burbs, The (1989)                 2.910891  1989
10 Things I Hate About You (1999)  3.422857  1999
101 Dalmatians (1961)              3.596460  1961
101 Dalmatians (1996)              3.046703  1996
12 Angry Men (1957)                4.295455  1957

+ We should add **"year"** column **:**
<br> **"year"** sütunu eklemeliyiz **:**

In [110]:
#   From 'data_reduced' select only title and year columns , drop duplicates and set title as index ::

year_info = data_reduced

In [111]:
#   Now assign this "year_info" to 'year' Column ::
#   Şimdi bu "year_info"'yu , 'year'Sütununa atayın ::

avg_rating_by_title["year"] = year_info

ValueError: Columns must be same length as key

In [112]:
avg_rating_by_title

rating  year
title                                            
'burbs, The (1989)                 2.910891  1989
10 Things I Hate About You (1999)  3.422857  1999
101 Dalmatians (1961)              3.596460  1961
101 Dalmatians (1996)              3.046703  1996
12 Angry Men (1957)                4.295455  1957
...                                     ...   ...
Young Guns (1988)                  3.418149  1988
Young Guns II (1990)               2.907859  1990
Young Sherlock Holmes (1985)       3.390501  1985
Zero Effect (1998)                 3.750831  1998
eXistenZ (1999)                    3.256098  1999

[1216 rows x 2 columns]

+ Write a function that takes a DataFrame and sorts and returns the first element **:**
<br> Bir Veri Çerçevesi alan ve ilk öğeyi sıralayan ve döndüren bir fonksiyon yazın **:**
+ Yani artık Yıllara ve Reytinglere göre bir analiz yapabiliriz... Mesela **, 1989** yılındaki verileri bulup **,** reytinge göre **sıralayıp en fazla reytinge sahip** olan filmi yazdırabiliriz **:**

In [113]:
#   Bir fonksiyon yazalım... Bu fonksiyon içerisine "avg_rating_by_title" DataFrame'sini yazdığımızda ,
# reytinge göre bir Sıralama yapsın ve en fazla reytinge sahip filmi döndürsün ::
#   Ve bunu da Index şeklinde döndürsün ::

def find_top_movie(x):
    return x.nlargest(1 , "rating").index[0]

+ Test this in **"avg_rating_by_title" :**
<br> Şimdi bunu **, "avg_rating_by_title"** DataFrame'mizin içerisinde test edelim **:**

In [114]:
find_top_movie(avg_rating_by_title)

'Seven Samurai (The Magnificent Seven) (Shichinin no samurai) (1954)'

+ Now apply **"find_top_movie" :**
<br> Şimdi **``groupby().apply()``**'da **"find_top_movie"**'yi uygulayalım **:**
+ Bu yarattığımız denklemi **,** TÜM YILLARDA uygulamasını istiyorsak **,** bunu yapmalıyız **:**

In [115]:
avg_rating_by_title.groupby("year").apply(find_top_movie)

year
1925                                Gold Rush, The (1925)
1926                                    Metropolis (1926)
1930                All Quiet on the Western Front (1930)
1931                                   City Lights (1931)
1933                                     Duck Soup (1933)
                              ...                        
1996    Wallace & Gromit: The Best of Aardman Animatio...
1997           Life Is Beautiful (La Vita è bella) (1997)
1998                           Saving Private Ryan (1998)
1999                              Sixth Sense, The (1999)
2000                                 Almost Famous (2000)
Length: 70, dtype: object

Yukarıda gördüğümüz gibi **,** HER YIL için EN ÇOK reytinge sahip filmleri bu şekilde bulabiliyoruz...

---
---

**---------->**

**++++ NOTE ++++**
<br> **GROUPBY** fonksiyonu **;** içerisine yazdığımız Kolonun HER ELEMANI için **yeni ayrı bir Tablo** yaratıyor... **APPLY** fonksiyonu da **;** içerisine yazdığımız denklemi **,** GROUPBY'ın içerisine yazdığımız Kolonun HER BİR elemanına uyguluyor... Yani aslında **,** GROUPBY'ın yarattığı HER DataFrame **,** APPLY fonksiyonunun içerisinde yazdığımız denklemin içine tek tek giriyor ve denklem **,** bu tablolara uygulatılıyor...

**<----------**

---
---

+ Now print the number of movie each year as well **:**
<br> Şimdi **, her yıl için toplam film sayısını** da yazdırın **:**
<br> Yani mesela **1925** yılında toplam kaç film yapılmış onu da göstersin **:**
<br> Bunu nasıl yapabiliriz **?** Amacımız **her yıldaki toplam eleman sayısını** bulmaktır... En basiti mesela **"``len( )``"** fonksiyonu kullanılabilir **:**

In [95]:
avg_rating_by_title.head()

rating  year
title                                            
'burbs, The (1989)                 2.910891  1989
10 Things I Hate About You (1999)  3.422857  1999
101 Dalmatians (1961)              3.596460  1961
101 Dalmatians (1996)              3.046703  1996
12 Angry Men (1957)                4.295455  1957

+ Yukarıda yaratmış olduğumuz denkleme "**len()**" fonksiyonunu da ekleyelim **:**
+ Ve daha güzel görünmesi için de SERİ biçiminde göstermesini isteyelim **:**
+ Ve daha da güzel görünmesi için SERİ'ye isim verelim **(index=...) :**

In [116]:
def find_top_movie(x):
    return pd.Series([x.nlargest(1, "rating").index[0] , len(x)] , index=["title", "count"])

In [117]:
find_top_movie(avg_rating_by_title)

title    Seven Samurai (The Magnificent Seven) (Shichin...
count                                                 1216
dtype: object

In [118]:
avg_rating_by_title.groupby("year").apply(find_top_movie)

title  count
year                                                          
1925                              Gold Rush, The (1925)      1
1926                                  Metropolis (1926)      1
1930              All Quiet on the Western Front (1930)      1
1931                                 City Lights (1931)      3
1933                                   Duck Soup (1933)      2
...                                                 ...    ...
1996  Wallace & Gromit: The Best of Aardman Animatio...     73
1997         Life Is Beautiful (La Vita è bella) (1997)     73
1998                         Saving Private Ryan (1998)     83
1999                            Sixth Sense, The (1999)    101
2000                               Almost Famous (2000)     50

[70 rows x 2 columns]

Yukarıda gördüğümüz gibi **,** mesela **1925** yılında **sadece 1 tane** film varmış...

+ Mesela şimdi de **,** direkt **en çok reytinge sahip filmin reytingini** bulmaya çalışalım **:**

In [119]:
avg_rating_by_title.nlargest(1 , "rating").rating.values

array([4.56050955])

Bu yukarıdaki satırda yazdığımız kodu kullanarak bir denklem yaratırsak **,** HER YIL İÇİN EN ÇOK reyting almış olan filmin **,** ne kadar reyting aldığını bulabiliriz...